File: backend/services/malya_service.py
"""
Malya module service
Business logic for Malya (Ideas/Creativity) module
"""
from datetime import datetime
from typing import List
from ..models.malya_models import MalyaIdea, CreateIdeaRequest, IdeasResponse
import uuid


class MalyaService:
    """Service for Malya module"""

    def __init__(self):
        # In-memory storage (stub)
        self.ideas: List[MalyaIdea] = []

    def get_ideas(self) -> IdeasResponse:
        """Get all ideas"""
        return IdeasResponse(
            ideas=self.ideas,
            count=len(self.ideas)
        )

    def create_idea(self, request: CreateIdeaRequest) -> MalyaIdea:
        """Create a new idea"""
        now = datetime.utcnow().isoformat()
        idea = MalyaIdea(
            id=str(uuid.uuid4()),
            title=request.title,
            description=request.description,
            category=request.category or "creative",
            status="draft",
            priority=request.priority or "medium",
            tags=request.tags,
            created_at=now,
            updated_at=now,
            related_modules=[]
        )
        self.ideas.append(idea)
        return idea

    def get_idea(self, idea_id: str) -> MalyaIdea:
        """Get idea by ID"""
        for idea in self.ideas:
            if idea.id == idea_id:
                return idea
        raise ValueError(f"Idea {idea_id} not found")

    def update_idea(self, idea_id: str, updates: dict) -> MalyaIdea:
        """Update an idea"""
        for i, idea in enumerate(self.ideas):
            if idea.id == idea_id:
                updates['updated_at'] = datetime.utcnow().isoformat()
                updated_idea = idea.copy(update=updates)
                self.ideas[i] = updated_idea
                return updated_idea
        raise ValueError(f"Idea {idea_id} not found")

    def delete_idea(self, idea_id: str) -> None:
        """Delete an idea"""
        self.ideas = [i for i in self.ideas if i.id != idea_id]


# Singleton instance
malya_service = MalyaService()

File: frontend/src/services/modules/malyaService.ts
/**
 * Malya Module API Service
 * Handles all API calls for Malya (Ideas) module
 */

import { apiClient } from '../api';

export interface MalyaIdea {
  id: number;
  title: string;
  description: string;
  idea_type?: string;
  status?: string;
  related_ideas?: number[];
  tags?: string[];
  time?: string;
  module?: string;
  source_trace?: string;
  canon_bundle_id?: string;
}

export interface MalyaIdeaCreate {
  title: string;
  description: string;
  idea_type?: string;
  status?: string;
  related_ideas?: number[];
  tags?: string[];
}

export interface MalyaIdeaUpdate {
  title?: string;
  description?: string;
  idea_type?: string;
  status?: string;
  related_ideas?: number[];
  tags?: string[];
}

const malyaService = {
  /**
   * Get Malya module status
   */
  async getStatus() {
    const response = await apiClient.get('/api/v1/malya/');
    return response.data;
  },

  /**
   * Create a new idea
   */
  async createIdea(idea: MalyaIdeaCreate): Promise<MalyaIdea> {
    const response = await apiClient.post('/api/v1/malya/ideas', idea);
    return response.data;
  },

  /**
   * Get all ideas with pagination
   */
  async getIdeas(params?: {
    skip?: number;
    limit?: number;
    idea_type?: string;
    status?: string;
  }): Promise<MalyaIdea[]> {
    const response = await apiClient.get('/api/v1/malya/ideas', { params });
    return response.data;
  },

  /**
   * Get idea by ID
   */
  async getIdea(id: number): Promise<MalyaIdea> {
    const response = await apiClient.get(`/api/v1/malya/ideas/${id}`);
    return response.data;
  },

  /**
   * Update idea
   */
  async updateIdea(id: number, updates: MalyaIdeaUpdate): Promise<MalyaIdea> {
    const response = await apiClient.put(`/api/v1/malya/ideas/${id}`, updates);
    return response.data;
  },

  /**
   * Delete idea
   */
  async deleteIdea(id: number): Promise<void> {
    await apiClient.delete(`/api/v1/malya/ideas/${id}`);
  },

  /**
   * Get ideas by status
   */
  async getIdeasByStatus(status: string): Promise<MalyaIdea[]> {
    return this.getIdeas({ status });
  },

  /**
   * Get active ideas (not archived/completed)
   */
  async getActiveIdeas(): Promise<MalyaIdea[]> {
    return this.getIdeas({ status: 'active' });
  },
};

export default malyaService;
